File: Array/subsets.py
# Problem Link :- https://leetcode.com/problems/subsets/

"""Given a set of distinct integers, nums, return all possible subsets (the power set)."""

class Solution(object):
    def subsets(self, nums):
        p = len(nums)-1
        n = 1 << len(nums)
        ans = []
        for i  in range(n):
            l = []
            for j in range(p+1):
                if i & (1 << j):
                    l.append(nums[p-j])
                else:
                    continue
            ans.append(l)
        return ans
                

File: BackTracking/Subsets.cpp
//Problem Link:- https://leetcode.com/problems/subsets/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void helper(vector<vector<int>>& result,vector<int>& path,vector<int>nums, int index){
        if (index >= nums.size()){
            result.push_back(path);
            return;
        }
        path.push_back(nums[index]);
        helper(result,path,nums,index+1);
        path.pop_back();
        helper(result,path,nums,index+1);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        if (nums.size() == 0)
            return result;
        vector<int> path;
        helper(result,path,nums,0);
        return result;
    }
};